File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        
        
        def binary_search(nums, target, is_searching_left):
            l = 0
            r = len(nums) - 1
            idx = -1
            
            while l <= r:
                mid = (l + r) // 2
                
                if nums[mid] > target:
                    r = mid - 1
                elif nums[mid] < target:
                    l = mid + 1
                else:
                    idx = mid
                    
                    if is_searching_left:
                        r = mid - 1
                    else:
                        l = mid + 1
            
            return idx
        
        left = binary_search(nums, target, 1)
        right = binary_search(nums, target, 0)
        
        return [left, right]

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int left = bin_search(nums, target, true);
        int right = bin_search(nums, target, false);
        vector<int>ans{left, right};
        return ans;
    }
    int bin_search(vector<int>& nums, int target, bool isleft){
        int l = 0, r = nums.size()-1;
        int idx = -1;
        while (l <= r){
            int mid = l+(r-l)/2;
            if (nums[mid] > target){
                r = mid-1;
            }
            else if (nums[mid] < target){
                l = mid+1;
            }
            else{
                idx = mid;
                if (isleft){
                    r = mid-1;
                }
                else{
                    l = mid+1;
                }
            }
        }
        return idx;
    }
};